cogs/audio/audio.py

```diff
@@ -1,112 +1,367 @@
+        self.task = self.loop.create_task(self.queue_manager())
+
+    def enqueue(self, source: discord.AudioSource):
+        self.queue.append(source)
 
+    def mix(self, source: discord.AudioSource):
+        # Check if voice client is already playing
+        if self.is_busy(self.voice_client):
+            self.voice_client.source.mix(source)
+        else:
+            self.voice_client.play(MixedSource(source))
+
+    def skip(self):
+        try:
+            if isinstance(self.voice_client.source.sources[0], YTDLSource):
+                self.voice_client.source.sources[0].cleanup()
+                self.voice_client.source.sources.pop(0)
+            else:
+                raise QueueError("I'm not playing any music.")
+        except (AttributeError, IndexError):
+            raise QueueError("I'm not playing anything.")
+
+    def restart(self):
+        if self.task.done():
+            self.disc_timer = 0
+            self.task = self.loop.create_task(self.queue_manager())
+
+    def set_voice_client(self, voice_client: discord.VoiceClient):
+        self.voice_client = voice_client
 
-# Just a little experimental audio cog not meant for final release
+    async def queue_manager(self):
+        while True:
+            await asyncio.sleep(1)
+            self.disc_timer += 1
+            if self.is_busy(self.voice_client):
+                self.disc_timer = 0
+            if self.disc_timer > self.timeout:
+                await self.voice_client.disconnect()
+                # This kills the task
+                return
+            if not self.queue:
+                continue
+            if self.queue[0].done:
+                # current song is done
+                self.queue.pop(0)
+            if self.voice_client is None:
+                # can put some reconnect logic here later
+                # or at least an error message
+                continue
+            if self.queue and not self.is_busy(self.voice_client):
+                # ok to play new song
+                self.voice_client.play(MixedSource(self.queue[0]))
+
+    def is_busy(self, voice_client: discord.VoiceClient):
+        return voice_client.is_playing() or voice_client.is_paused()
 
 
 class Audio:
     """Audio commands"""
 
+    default_global_settings = {
+
+    }
+
+    default_guild_settings = {
+
+    }
+
     def __init__(self, bot):
         self.bot = bot
+        self.conf = Config.get_conf(self, identifier=8675309)
+        self.conf.register_global(
+            **self.default_global_settings
+        )
+        self.conf.register_guild(
+            **self.default_guild_settings
+        )
+        self.data_path = cog_data_path(self)
+        self.local_path = os.path.join(self.data_path, 'local')
+        self.downloader_path = os.path.join(self.data_path, 'cache')
+        self.downloader = Downloader(self.downloader_path)
+        if not os.path.exists(self.local_path):
+            os.makedirs(self.local_path)
```

Comment:
`pathlib` will also make this easier, turning it into `self.local_path.mkdir(exists_ok=True)`

Reply:
Done and done! Only problem I ran into with pathlib was that FFmpegPCMAudio did not enjoy being fed the Path object. Using the string representation worked.